File: XMNLaunchViewExample/XMNLaunchView.m
//
//  XMNLaunchView.m
//  XMNLaunchViewExample
//
//  Created by XMFraker on 16/11/9.
//  Copyright © 2016年 XMFraker. All rights reserved.
//

#import "XMNLaunchView.h"
#import "YYAnimatedImageView.h"

#import "YYWebImage.h"

#import "UIImage+YYWebImage.h"
#import "UIImageView+YYWebImage.h"


NSInteger kXMNLaunchViewTag = 12306;

@interface XMNLaunchView ()

@property (weak, nonatomic) YYAnimatedImageView *imageView;
@property (weak, nonatomic) UIButton *skipButton;
@property (strong, nonatomic) YYWebImageManager *imageManager;

@property (strong, nonatomic, nullable) NSURL   *imageURL;


@property (strong, nonatomic) NSTimer *countdownTimer;
@property (strong, nonatomic) NSDate *countdownDate;

@end

@implementation XMNLaunchView

#pragma mark - Life Cycle

- (instancetype)initWithPlaceholder:(UIImage *)placeholder
                           imageURL:(NSURL *)imageURL {
    
    if (self = [super initWithFrame:[UIApplication sharedApplication].keyWindow.bounds]) {
        
        NSOperationQueue *queue = [[NSOperationQueue alloc] init];
        queue.name = @"com.XMFraker.XMNLaunchView.kXMNDownloadLaunchQueue";
        if ([queue respondsToSelector:@selector(setQualityOfService:)]) {
            queue.qualityOfService = NSQualityOfServiceBackground;
        }
        self.tag = kXMNLaunchViewTag;
        
        self.imageManager = [[YYWebImageManager alloc] initWithCache:[YYImageCache sharedCache] queue:queue];
        
        self.autoHide = YES;
        self.displayTimeout = 3.f;
        self.requestTimeout = 5.f;
        self.placeholder = placeholder ? : [XMNLaunchView launchImage];
        
        [self setupUI];
        
        [self launchImageWithURL:imageURL
                  requestTimeout:self.requestTimeout
                  displayTimeout:self.displayTimeout];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    
    return [self initWithPlaceholder:nil
                            imageURL:nil];
}

- (instancetype)initWithFrame:(CGRect)frame {
    
    return [self initWithPlaceholder:nil
                            imageURL:nil];
}

#pragma mark - Method

- (void)setupUI {
    
    YYAnimatedImageView *animatedImageView = [[YYAnimatedImageView alloc] initWithFrame:self.bounds];
    animatedImageView.autoPlayAnimatedImage = YES;
    animatedImageView.userInteractionEnabled = NO;
    [self addSubview:self.imageView = animatedImageView];
    
    UITapGestureRecognizer *tapGes = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(handleTapLaunch)];
    [animatedImageView addGestureRecognizer:tapGes];
    
    UIButton *skipButton = [UIButton buttonWithType:UIButtonTypeCustom];
    [skipButton setTitle:[NSString stringWithFormat:@"%02ds",(int)self.displayTimeout] forState:UIControlStateNormal];
    skipButton.layer.cornerRadius = 15.f;
    skipButton.frame = CGRectMake(self.bounds.size.width - 16 - 60, 32, 60, 30);
    skipButton.backgroundColor = [UIColor colorWithRed:0.f green:0.f blue:0.f alpha:.5f];
    [skipButton.titleLabel setFont:[UIFont systemFontOfSize:12.f]];
    [skipButton addTarget:self action:@selector(handleSkipAction:) forControlEvents:UIControlEventTouchUpInside];
    skipButton.hidden = YES;
    [self addSubview:self.skipButton = skipButton];
}

- (void)launchImageWithURL:(NSURL *)imageURL
            requestTimeout:(NSInteger)requestTimeout
            displayTimeout:(NSInteger)displayTimeout {
    
    self.skipButton.hidden = YES;
    self.requestTimeout = requestTimeout > 0 ? requestTimeout : 5.f;
    self.displayTimeout = displayTimeout > 0 ? displayTimeout : 3.f;
    
    self.imageView.image = self.placeholder;
    self.imageURL = imageURL;
    if (imageURL) {
        
        __weak typeof(*&self) wSelf = self;
        [self.imageView yy_cancelCurrentImageRequest];
        [self.imageView yy_setImageWithURL:imageURL placeholder:self.placeholder options:YYWebImageOptionSetImageWithFadeAnimation manager:self.imageManager progress:nil transform:^UIImage * _Nullable(UIImage * _Nonnull image, NSURL * _Nonnull url) {
            
            YYImage *animatedImage = (YYImage *)image;
            if (animatedImage.animatedImageType == YYImageTypeGIF) {
                /** gif图片不做处理 */
                return animatedImage;
            }else {
                return [image yy_imageByResizeToSize:[UIApplication sharedApplication].keyWindow.bounds.size contentMode:UIViewContentModeScaleAspectFill];
            }
            /** 适配图片 */
        } completion:^(UIImage * _Nullable image, NSURL * _Nonnull url, YYWebImageFromType from, YYWebImageStage stage, NSError * _Nullable error) {
            
            __strong typeof(*&wSelf) self = wSelf;
            if (image && !error) {
                
                /** 显示跳过按钮,开始显示倒计时 */
                self.imageView.userInteractionEnabled = YES;
                self.skipButton.hidden = NO;
                self.countdownDate = [NSDate date];
                self.countdownTimer = [NSTimer scheduledTimerWithTimeInterval:1.f
                                                                       target:self
                                                                     selector:@selector(handleCountdownAction)
                                                                     userInfo:nil
                                                                      repeats:YES];
            }else if (self.isAutoHide){
                
                [self dismissLaunchViewWithMode:XMNLaunchViewDismissModeDefault];
            }
        }];
    }
}

- (void)dismissLaunchView {
    
    [self dismissLaunchViewWithMode:XMNLaunchViewDismissModeDefault];
}

- (void)dismissLaunchViewWithMode:(XMNLaunchViewDismissMode)mode {
    
    self.countdownTimer ? [self.countdownTimer invalidate] : nil;
    self.skipButton.hidden = YES;
    
    [UIView animateWithDuration:.5f animations:^{
        
        self.alpha = 0.3f;
        self.transform = CGAffineTransformMakeScale(1.5f, 1.5f);
    } completion:^(BOOL finished) {
        
        switch (mode) {
            case XMNLaunchViewDismissModeTap:
                self.completionBlock ? self.completionBlock(self, mode) : nil;
            default:
                [self removeFromSuperview];
                break;
        }
    }];
}

#pragma mark - Events

- (void)handleSkipAction:(UIButton *)button {
    
    [self dismissLaunchViewWithMode:XMNLaunchViewDismissModeDefault];
}

- (void)handleTapLaunch {
    
    [self dismissLaunchViewWithMode:XMNLaunchViewDismissModeTap];
}

- (void)handleCountdownAction {
    
    NSTimeInterval remainTimeInveral = MAX((int)self.displayTimeout - (int)[[NSDate date] timeIntervalSinceDate:self.countdownDate], 0);
    [self.skipButton setTitle:[NSString stringWithFormat:@"%02ds",(int)remainTimeInveral] forState:UIControlStateNormal];
    
    if (remainTimeInveral <= 0) {
        [self dismissLaunchViewWithMode:XMNLaunchViewDismissModeDefault];
    }
}

#pragma mark - Setter

- (void)setPlaceholder:(UIImage *)placeholder {
    
    _placeholder = placeholder;
    self.imageView.image = placeholder;
}

- (void)setDisplayTimeout:(NSTimeInterval)displayTimeout {
    
    _displayTimeout = displayTimeout;
    [self.skipButton setTitle:[NSString stringWithFormat:@"%02ds",(int)self.displayTimeout] forState:UIControlStateNormal];
}

- (void)setRequestTimeout:(NSTimeInterval)requestTimeout {
    
    self.imageManager.timeout = requestTimeout;
}

- (void)setFrame:(CGRect)frame {
    
    [super setFrame:frame];
    self.imageView.frame = self.bounds;
}

#pragma mark - Getter

- (NSTimeInterval)requestTimeout {
    
    return self.imageManager.timeout;
}

- (BOOL)isAutoHide {
    
    return _autoHide;
}

#pragma mark - Class Method


/**
 从启动图中获取当前启动图片
 
 @return 获取到的图片
 */
+ (UIImage * __nullable)launchImage {
    
    NSString *viewOrientation;
    switch ([UIApplication sharedApplication].statusBarOrientation) {
        case UIInterfaceOrientationLandscapeLeft:
        case UIInterfaceOrientationLandscapeRight:
            viewOrientation = @"Landscape";
            break;
        case UIInterfaceOrientationPortrait:
        case UIInterfaceOrientationPortraitUpsideDown:
        default:
            viewOrientation = @"Portrait";
            break;
    }
    NSString *launchImageName = nil;
    NSArray* imagesDict = [[[NSBundle mainBundle] infoDictionary] valueForKey:@"UILaunchImages"];
    for (NSDictionary* dict in imagesDict) {
        CGSize imageSize = CGSizeFromString(dict[@"UILaunchImageSize"]);
        if (CGSizeEqualToSize(imageSize, [UIApplication sharedApplication].keyWindow.bounds.size) && [viewOrientation isEqualToString:dict[@"UILaunchImageOrientation"]]) {
            launchImageName = dict[@"UILaunchImageName"];
        }
    }
    return launchImageName ? [UIImage imageNamed:launchImageName] : nil;
}


+ (XMNLaunchView * __nullable)launchViewOnWindow {
    
    XMNLaunchView *view = [[UIApplication sharedApplication].keyWindow viewWithTag:kXMNLaunchViewTag];
    if ([view isKindOfClass:[XMNLaunchView class]]) {
        return view;
    }
    return nil;
}
@end

File: XMNLaunchViewExample/XMNLaunchView.swift
//
//  XMNLaunchView.swift
//  XMNLaunchViewExample
//
//  Created by XMFraker on 16/8/23.
//  Copyright © 2016年 XMFraker. All rights reserved.
//


import UIKit

public enum XMNLaunchViewDismissMode: Int {
    
    /** 用户点击广告消失 */
    case tap
    /** 用户点击跳转消失 */
    case skip
    /** 图片展示时间到了 */
    case displayOverTime
    /** 广告超时消失,获取图片超时消失 */
    case requestOverTime
    /** 未设置imageURL,获取请求图片失败 */
    case noneImage
}

class XMNLaunchView: UIView {
    
    /** 回调block */
    fileprivate var completedBlock: ((_ dismissmode: XMNLaunchViewDismissMode) -> Void)?

    weak var displayWindow: UIWindow?
    
    var imageURL: URL? {
        
        //需要注意的是didSet,willSet 在类初始化的时候并不会调用
        //需要使用KVC模式 才会调用
        didSet {
            self.displayImage()
        }
    }
    
    var startDate: Date? {
        didSet {
            if self.startDate != nil {
                //停止之前的timer  否则不开启tiemr,因为没有timer 说明还没开始显示图片
                if self.timer != nil {
                    self.timer?.invalidate()
                    //重启一个timer
                    self.timer = Timer.scheduledTimer(timeInterval: 1.0, target: self, selector: #selector(XMNLaunchView.handleTimerAction(_:)), userInfo: nil, repeats: true)
                }
            }else {
                if self.timer != nil {
                    self.timer?.invalidate()
                }
            }
        }
    }
    var timer: Timer?
    
    /**
     使用懒加载方法,初始化imageView
     
     - returns:
     */
    lazy var imageView:YYAnimatedImageView = {
        
        let imageView = YYAnimatedImageView(frame: self.displayWindow!.bounds)
        imageView.contentMode = .scaleAspectFill
        imageView.isUserInteractionEnabled = true
        imageView.alpha = 0.0
        
        let tapGes = UITapGestureRecognizer(target: self, action: #selector(XMNLaunchView.handleTapAction(_:)))
        imageView.addGestureRecognizer(tapGes)
        return imageView
    }()
    
    
    /**
     使用懒加载方法,生成skipButton
     
     - returns:
     */
    lazy var skipButton:UIButton = {
       
        let button = UIButton(type: .custom)
        button.frame = CGRect(x: self.bounds.size.width - 16 - 60, y: 32, width: 60, height: 30)
        button.setTitle("\(self.displayInterval)s", for: UIControlState())
        
        let backgroundColor = UIColor(red: 0.0, green: 0.0, blue: 0.0, alpha: 0.3)
        let backgroundImage = UIImage.yy_image(with: backgroundColor, size: CGSize(width: 60, height: 30))?.yy_image(byRoundCornerRadius: 15.0)
        button.addTarget(self, action: #selector(handleButtonAction(_:)), for: .touchUpInside)
        button.setBackgroundImage(backgroundImage, for: UIControlState())
        return button
    }()

    /// 显示 图片的时间
    var displayInterval: Int {
        
        didSet {
            //修改显示时间,则重启timer
            self.startDate = Date()
        }
    }
    var requestInterval: Int {
        
        didSet {
            //修改了requestInterval 重新设置超时方法
            NSObject.cancelPreviousPerformRequests(withTarget: self, selector: #selector(XMNLaunchView.handleRequestTimeOutAction), object: nil)
            self.perform(#selector(XMNLaunchView.handleRequestTimeOutAction), with: nil, afterDelay: Double(self.requestInterval))
        }
    }
    
    var launchImage: UIImage? {
        var launchImageName: String? = nil
        if let imageDicts = Bundle.main.infoDictionary?["UILaunchImages"] as? Array<AnyObject> {
            for imageDict in imageDicts{
                if let dict = imageDict as? NSDictionary {
                    let size = CGSizeFromString(dict["UILaunchImageSize"] as! String)
                    if size.equalTo((self.displayWindow?.bounds.size)!) {
                        launchImageName = dict["UILaunchImageName"] as? String
                    }
                }
                print(imageDict)
            }
            if let imageName = launchImageName {
                return UIImage(named: imageName)
            }else {
                return nil
            }
        }
        return nil
    }
    

    init(displayWindow: UIWindow, imageURL: URL?) {
        
        self.displayInterval = 3
        self.requestInterval = 10
        super.init(frame: displayWindow.bounds)
        self.displayWindow = displayWindow
        self.imageURL = imageURL

        self.setupUI()
        self.displayImage()
        
        //首次初始化 手动开启超时
        self.perform(#selector(XMNLaunchView.handleRequestTimeOutAction), with: nil, afterDelay: Double(self.requestInterval))
    }
    
    required convenience init?(coder aDecoder: NSCoder) {
        
        self.init(displayWindow: UIApplication.shared.keyWindow!, imageURL: nil)
    }
    
    /**
     重写析构方法,释放timer
     */
    deinit {
        if self.timer != nil {
            self.timer?.invalidate()
        }
        NSObject.cancelPreviousPerformRequests(withTarget: self, selector: #selector(XMNLaunchView.handleRequestTimeOutAction), object: nil)
        print("i am deinit \(NSStringFromClass(self.classForCoder))")
    }

    // MARK: Method
    
    /**
     设置完成回调
     
     - parameter block: 
     */
    func setCompletedBlock(_ block : @escaping (_: XMNLaunchViewDismissMode) -> Void) -> Void {
        self.completedBlock = block
    }
    
    func setupUI() {

        if self.launchImage != nil {
            self.backgroundColor = UIColor(patternImage: self.launchImage!)
        }else {
            self.backgroundColor = UIColor.white
        }
        //注意此处调用顺序, 先将keyWindow显示出来,再添加launchView, 防止launchView被添加到displayWindow.rootViewController.view后面
        self.displayWindow?.makeKeyAndVisible()
        self.displayWindow?.addSubview(self)
    }
    
    /**
     显示图片,如果没有imageURL,则直接隐藏launchView
     */
    func displayImage() {
        
        if self.imageURL != nil {
            
            if self.subviews.contains(self.imageView) {
                self.imageView.alpha = 0.0
            }else {
                self.addSubview(self.imageView)
            }
            
            self.imageView.yy_setImage(with: self.imageURL, placeholder: nil, options: .avoidSetImage, completion: { [unowned self] (downloadImage, originURL, fromType, imageState, error) in
                
                //判断下载的图片是否是YYImage
                if error != nil {
                    self.dismissLaunchView(dismissMode: .noneImage)
                    return
                }
                
                //取消请求超时的方法执行,已经获取到方法了
                NSObject.cancelPreviousPerformRequests(withTarget: self, selector: #selector(XMNLaunchView.handleRequestTimeOutAction), object: nil)

                if let image = downloadImage as? YYImage {
                    if image.animatedImageType == .GIF {
                        
                        self.imageView.animatedImage = image
                        self.imageView.startAnimating()
                    }else {
                        self.imageView.image = image.yy_imageByResize(to: (self.displayWindow?.bounds.size)!)
                    }
                    self.imageView.alpha = 0.0
                    UIView.animate(withDuration: 0.5, animations: {
                        self.imageView.alpha = 1.0
                        }, completion: { (_) in
                            self.startDate = Date()
                            self.timer = Timer.scheduledTimer(timeInterval: 1.0, target: self, selector: #selector(XMNLaunchView.handleTimerAction(_:)), userInfo: nil, repeats: true)
                            self.addSubview(self.skipButton)
                            self.backgroundColor = UIColor.clear
                    })
                }
            })
        }else {
         
            self.dismissLaunchView(dismissMode: .noneImage)
        }
    }
    
    /**
     处理timer方法
     
     - parameter timer:
     */
    func handleTimerAction(_ timer: Timer) {
        
        let remainTimeInterval = max(self.displayInterval - Int(Date().timeIntervalSince(self.startDate!)), 0)
        self.skipButton.setTitle("\(remainTimeInterval)s", for: UIControlState())
        if remainTimeInterval <= 0 {

            self.dismissLaunchView(dismissMode: .displayOverTime)
        }
        print("timer action")
    }
    
    /**
     处理请求超时
     */
    func handleRequestTimeOutAction() {

        self.dismissLaunchView(dismissMode: .requestOverTime)
    }
    
    /**
     隐藏launchView
     
     - parameter mode:
     */
    func dismissLaunchView(dismissMode mode :XMNLaunchViewDismissMode)  {
        
        NSObject.cancelPreviousPerformRequests(withTarget: self, selector: #selector(XMNLaunchView.handleRequestTimeOutAction), object: nil)
        
        if self.timer != nil {
            self.timer?.invalidate()
        }

        //只有此三种模式 skipButton 会被初始化显示出来,其他模式 不必去隐藏skipButton
        switch mode {
        case .displayOverTime: fallthrough
        case .skip: fallthrough
        case .tap : self.skipButton.isHidden = true
        default:
            break
        }
        
        UIView.animate(withDuration: 0.5, animations: { 
                self.imageView.alpha = 0.3
                self.imageView.transform = CGAffineTransform(scaleX: 1.5, y: 1.5)
            }, completion: { (_) in
                
                if let block = self.completedBlock {
                    block(mode)
                }
                self.removeFromSuperview()
        }) 
    }
    
    /**
     处理imageView的点击手势
     
     - parameter gesture:
     */
    func handleTapAction(_ gesture: UITapGestureRecognizer) {
        
        print("get tap action")
    }
    
    /**
     处理skipButton 跳过功能
     
     - parameter button:
     */
    func handleButtonAction(_ button: UIButton) {
        
        self.dismissLaunchView(dismissMode: .skip)
    }
    
}
